### ISSUE TITLE
Best practices for using sudo with Sensu checks? 

### ISSUE THREAD

**Issue by calebhailey** [sensu/sensu-docs#386]

**Comment by GhostLyrics** [comment]
Since I needed to use some checks and metrics which required the use of `sudo` - which was better than adding the `sensu` user to loads and loads of different groups, implicitly granting write permissions to files - I came up with the following snippet for use in `/etc/sudoers.d/sensu`.

```
# allow using only the binary name instead of the full path in check definitions
Defaults:sensu secure_path="/opt/sensu/embedded/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"

# blindly allow all sensu-install provided checks and metrics which follow the proper naming convention with all possible arguments
sensu ALL=(ALL) NOPASSWD: /opt/sensu/embedded/bin/check*.rb *
sensu ALL=(ALL) NOPASSWD: /opt/sensu/embedded/bin/metrics*.rb *
```

Inspiration for this kind of usage came after reading:
- [Michael W Lucas: Sudo: You're Doing It Wrong, BSDCcan 2014](https://www.bsdcan.org/2014/schedule/attachments/283_2014-04-29%20sudo%20tutorial%20-%20bsdcan%202014.pdf)
- [Stefan Horlacher: Dangerous Sudoers Entries - PART 4: Wildcards](http://blog.csnc.ch/2012/10/dangerous-sudoers-entries-part-4-wildcards/)
- [serverfault: using wildcards for paths in sudoers file](http://serverfault.com/questions/383024/using-wildcards-for-paths-in-sudoers-file)

**Comment by majormoses** [comment]
I generally take minimal and explicit security approach where possible, while the above works it gives way too  much access IMHO. The checks themselves should never be run with sudo itself. It should spawn a subshell to elevate itself and then drop permissions. These commands need to be whitelisted in `/etc/sudoers.d/`. Say you need to remove a particular file you have now granted the use of `rm -Rf /` which would be catastrophic. The best solution is to always give access to the minimal required resources. Rather than directly calling `sudo rm $PATH_TO_FILE_OWNED_BY_ROOT` you should create a shell wrapper called something like `elevates-privs-and-removes-$SOME_FILE` and then allow this command to be executed in sudoers. This prevents argument expansion to be used nefariously. While this is an extreme example it demonstrates the danger of doing so. This is more important when you consider that scripts might allow an `eval` which would allow you to execute arbitrary command to create a new script and execute, as you have whitelisted the actual check script this will allow them to carry out a privilege escalation which could have been prevented.

**Comment by majormoses** [comment]
A less extreme example would be in the case of auto remediation, if a service is not running the logical thing to do is to attempt to start it. As the process is not likely owned by sensu this will require elevating privileges. While it might be tempting to just whitelist the service commands such as `/usr/bin/sv`, `/usr/bin/service'`, `/usr/sbin/service`, `/bin/systemd`, etc this is not recommended for the same reasons as above. You should create a per service wrapper that allows only starting of the service. This prevents checks from stopping either the intended or otherwise services.

**Comment by majormoses** [comment]
> This example has been developed with `safe_mode = true` in mind.

This should honestly be the default as it is terrifying in the context of allowing http apis for check creation, execution, scheduling, etc especially considering that most people do not understand how to properly secure systems.

**Comment by GhostLyrics** [comment]
@majormoses I tend to agree with you. What I dislike here is that you assume that everyone will have the additional time for engineering for developing and maintaining such a wrapper for every check that needs it.

**Comment by majormoses** [comment]
> What I dislike here is that you assume that everyone will have the additional time for engineering for developing and maintaining such a wrapper for every check that needs it.

I apologize if you felt like I was attacking you, that was not my intention. I used the :heart: emoji showing that I did like the solution. That being said I had a few extra thoughts as a security conscious engineer and based on the topic felt like they needed to be called out.

The title of this article is "Best Practices for using sudo with Sensu checks?" not "The Unfortunate Reality for using sudo with Sensu checks?". As much as I wish that I followed everything I said here, I don't but the goal was to discuss best practices. Even if you are aware of the risks and have accepted them, this is not something I would encourage people to do without explaining the consequences which I outlined. Many I would wager would not have noticed the vulnerabilities pointed out.
